server/ec2spotmanager/CloudProvider/EC2SpotCloudProvider.py

```diff
@@ -0,0 +1,299 @@
+
+
+class EC2SpotCloudProvider(CloudProvider):
+    def __init__(self):
+        self.logger = logging.getLogger("ec2spotmanager")
+
+    def terminate_instances(self, instances_ids_by_region):
+
+        for region in instances_ids_by_region:
+            cluster = EC2Manager(None)
+            try:
+                cluster.connect(region=region, aws_access_key_id=settings.AWS_ACCESS_KEY_ID,
+                                aws_secret_access_key=settings.AWS_SECRET_ACCESS_KEY)
+            except Exception as msg:
+                raise Exception({"type": "unclassified", "data": msg})
+
+            try:
+                self.logger.info("Terminating %s instances in region %s",
+                                 len(instances_ids_by_region[region]), region)
+                boto_instances = cluster.find(instance_ids=instances_ids_by_region[region])
+                # Data consistency checks
+                for boto_instance in boto_instances:
+                    # state_code is a 16-bit value where the high byte is
+                    # an opaque internal value and should be ignored.
+                    state_code = boto_instance.state_code & 255
+                    if not ((boto_instance.id in instances_ids_by_region[region]) or
+                            (state_code == INSTANCE_STATE['shutting-down'] or
+                             state_code == INSTANCE_STATE['terminated'])):
+                        self.logger.error("Instance with EC2 ID %s (status %d) "
+                                          "is not in region list for region %s",
+                                          boto_instance.id, state_code, region)
+
+                    cluster.terminate(boto_instances)
+            except (boto.exception.EC2ResponseError, boto.exception.BotoServerError, ssl.SSLError, socket.error) as msg:
+                self.logger.exception("terminate_instances: boto failure: %s", msg)
+                return
+
+    def start_instances(self, config, region, zone, userdata, image, instance_type, count=1):
+        images = self._create_laniakea_images(config)
+
+        # convert count from cores to instances
+        #
+        # if we have chosen the smallest possible instance that will put us over the requested core count,
+        #   we will only be spawning 1 instance
+        #
+        # otherwise there may be a remainder if this is not an even division. let that be handled in the next tick
+        #   so that the next smallest instance will be considered
+        #
+        # eg. need 12 cores, and allow instances sizes of 4 and 8 cores,
+        #     8-core instance costs $0.24 ($0.03/core)
+        #     4-core instance costs $0.16 ($0.04/core)
+        #
+        #     -> we will only request 1x 8-core instance this time around, leaving the required count at 4
+        #     -> next time around, we will request 1x 4-core instance
+        count = max(1, count // CORES_PER_INSTANCE[instance_type])
```

Comment:
this belongs in tasks.py, this method already takes the # of instances to start.